File: backend/app/routers/workspaces.py
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlmodel import Session

from app.database import get_session
from app.models.workspace import Workspace
from app.services import workspaces as svc

router = APIRouter(prefix="/workspaces", tags=["workspaces"])

SessionDep = Annotated[Session, Depends(get_session)]


class CreateWorkspaceRequest(BaseModel):
    name: str


@router.get("", response_model=list[Workspace])
def list_workspaces(session: SessionDep):
    return svc.list_workspaces(session)


@router.post("", response_model=Workspace, status_code=status.HTTP_201_CREATED)
def create_workspace(body: CreateWorkspaceRequest, session: SessionDep):
    return svc.create_workspace(session, body.name)


@router.get("/{id}", response_model=Workspace)
def get_workspace(id: uuid.UUID, session: SessionDep):
    workspace = svc.get_workspace(session, id)
    if workspace is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Workspace not found")
    return workspace


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_workspace(id: uuid.UUID, session: SessionDep):
    if not svc.delete_workspace(session, id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Workspace not found")

File: frontend/src/api/workspaces.ts
export interface Workspace {
  id: string
  name: string
}

export async function listWorkspaces(): Promise<Workspace[]> {
  const res = await fetch('/api/workspaces')
  if (!res.ok) throw new Error('Failed to fetch workspaces')
  return res.json()
}

export async function createWorkspace(name: string): Promise<Workspace> {
  const res = await fetch('/api/workspaces', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) throw new Error('Failed to create workspace')
  return res.json()
}

export async function getWorkspace(id: string): Promise<Workspace> {
  const res = await fetch(`/api/workspaces/${id}`)
  if (res.status === 404) throw new Error('Workspace not found')
  if (!res.ok) throw new Error('Failed to fetch workspace')
  return res.json()
}

export async function deleteWorkspace(id: string): Promise<void> {
  const res = await fetch(`/api/workspaces/${id}`, { method: 'DELETE' })
  if (res.status === 404) return
  if (!res.ok) throw new Error('Failed to delete workspace')
}
